### DISCUSSION TITLE
Filter collection by field on multiple relation field.

### DISCUSSION THREAD

**Discussion by Abyssopelagic** [pocketbase/pocketbase#5529] [Q&A] [answered]
I have a 'photos' collection with a tags field. Tags is a multiple relation field. The tags collection 'photo_tags' then has a person field. This field is a single relation to a user in 'users' collection. I want to be able to filter the photos collection given any amount of users. Like if i want to find a photo with these 3 people in it. This is being filtered by tags and the person related to it. I've tried both of these filters using `?=` and `?~`.

```javascript
filter: "tags.person.id ?= 'v7jqp9i04pdv2vf' && tags.person.id ?= 'vrstdmgbqadj6bf'"
filter: "tags.person ?= 'v7jqp9i04pdv2vf' && tags.person ?= 'vrstdmgbqadj6bf'"
```

If I filter by either person individually it works correctly, so these both work.
```javascript
filter: "tags.person.id ?= 'v7jqp9i04pdv2vf'"
filter: "tags.person.id ?= 'vrstdmgbqadj6bf'"
```

But when trying more than one the query returns empty.

Full call being
```javascript
pb.collection('photos').getList(page, perPage, {
    filter: "tags.person.id ?= 'v7jqp9i04pdv2vf' && tags.person.id ?= 'vrstdmgbqadj6bf'",
    expand: "tags",
    sort: "-date"
})
```

I've added expand for `tags.person` still no luck

Any suggestions on why this does not work?

**Comment by ganigeorgiev** [comment, marked as answer]
This is similar to https://github.com/pocketbase/pocketbase/discussions/5223, https://github.com/pocketbase/pocketbase/discussions/4319 and probably others.

The reason why it doesn't work is because rules evaluate in a single run and all `tags.person ?= x`  expressions targets the same row. It will generate a pseudo-query like:

```sql
SELECT photos.*
FROM photos
LEFT JOIN photo_tags tags ON photo_tags.id in photos.tags -- simplified fake expression
WHERE tags.person = 'v7jqp9i04pdv2vf' AND tags.person = 'vrstdmgbqadj6bf' -- cannot be 2 different values at the same time
```

There is no easy solution at the moment if you want to target client-side **both** at the same time.

If you want to return photos with tags related to any of the targeted persons you can replace `&&` with `||`.

If you want photos with tags related to both 2 listed persons, then in general, without changing your current collections structure, I can think of 2 options:
1. Send separate requests - find first the related tags for each person and then search for them in the photos:
    ```js
    const personIds = ["v7jqp9i04pdv2vf", "vrstdmgbqadj6bf"]

    // fetch all tags related to the targeted persons
    const tags = await pb.collection("photo_tags").getFullList({
        // filter: "person='id1' || person='id2' || ..."
        filter: personIds.map((id) => `person='${id}'`).join('||')
    })

    // generate photos "tags" filter for each person
    // [
    //    `(tags ~ '"person1TagId1"' || tags ~ '"person1TagId2"' || tags ~ '"person1TagId3"')`
    //    `(tags ~ '"person2TagId1"' || tags ~ '"person2TagId2"')`
    //    ...
    // ]
    perPersonTagsFilters = [];
    for (const personId of personIds) {
        const filter = tags.filter((tag) => tag.person = personId)
            .map((tag) => `tags ~ '"${tag.id}"'`)
            .join("||")
        perPersonTagsFilters.push("(" + filter + ")") // append the filter with parenthesis
    }
    
    // the tags filter group should equal the number of persons
    if (perPersonTagsFilters.length != personIds.length) {
        throw new Error("No photos tagged with all " + personIds.length " persons.")
    }

    const photos = await pb.collection("photos").getList(page, perPage, {
        // (tags ~ '"person1TagId1"' || tags ~ '"person1TagId2"' || tags ~ '"person1TagId3"') && (tags ~ '"person2TagId1"' || tags ~ '"person2TagId2"')
        filter: perPersonTagsFilters.join("&&")
    })
    ```

2. Create a helper [View collection](https://pocketbase.io/docs/collections/#view-collection) that will serialize the tags.person for example using [`json_group_array`](https://www.sqlite.org/json1.html#jgrouparray) and then you can join that collection via [back relation](https://pocketbase.io/docs/working-with-relations/#back-relations) and search for the persons with `?~` operator.

    Example `photoPersons` View collection query (_I haven't tested it_):
    ```sql
    SELECT
        photos.id,
        photos.id as photo, -- used for the back-relation
        json_group_array(tags.person) as persons -- should result in a json array with all the person ids
    FROM photos
    INNER JOIN json_each(photos.tags) je
    INNER JOIN photo_tags tags ON tags.id = je.value
    GROUP BY photos.id
    ```

    Then on the client side you can use it like:
    ```js
    const personIds = ["v7jqp9i04pdv2vf", "vrstdmgbqadj6bf"]

    const photos = await pb.collection("photos").getList(page, perPage, {
        // photoPersons_via_photo.persons ?~ '"id1"' && photoPersons_via_photo.persons ?~ '"id2"' && ...
        filter: personIds.map((id) => `photoPersons_via_photo.persons ?~ '"${id}"'`).join("&&")
    })
    ```

I haven't tested neither of the above 2 suggestions and in terms of performance it is difficult to say which one will perform better (_the first approach with the separate queries is more "safer" as it doesn't introduce extra joins which could affect the performance in large datasets_).

**Reply by Abyssopelagic** [reply]
Thanks for the quick reply! I had assumed that the filter ran for a single row at a time, just wasn't sure how to handle that. I went with the view_collection suggestion because I have some other query logic that would have become pretty complex with multiple requests being added in. Yes, #4319 is pretty similar, though I would not have known what to do with the extra relation step in my scenario. Sorry for the related question, but thanks again for the quick response!
